manifests/instance/tls.pp

```diff
@@ -0,0 +1,180 @@
+# @summary Configure TLS for an instance
+#
+# Requires LDAPI to be enabled
+#
+# @author https://github.com/simp/pupmod-simp-ds389/graphs/contributors
+#
+define ds389::instance::tls (
+  String[2]                                 $root_dn,
+  Stdlib::Absolutepath                      $root_pw_file,
+  Variant[Boolean, Enum['disabled','simp']] $ensure        = simplib::lookup('simp_options::pki', { 'default_value' => false}),
+  Simplib::Port                             $port          = 636,
+  String[1]                                 $source        = simplib::lookup('simp_options::pki', { 'default_value' => "/etc/pki/simp/${module_name}_${title}"}),
+  Stdlib::Absolutepath                      $cert          = "/etc/pki/simp_apps/${module_name}_${title}/x509/public/${facts['fqdn']}.pub",
+  Stdlib::Absolutepath                      $key           = "/etc/pki/simp_apps/${module_name}_${title}/x509/private/${facts['fqdn']}.pem",
+  Stdlib::Absolutepath                      $cafile        = "/etc/pki/simp_apps/${module_name}_${title}/x509/cacerts/cacerts.pem",
+  Ds389::ConfigItems                        $dse_config    = simplib::dlookup('ds389::instance::pki', 'dse_config', { 'default_value' => {} }),
+  String[16]                                $token         = simplib::passgen("ds389_${title}_pki", { 'length' => 32 }),
+  String[1]                                 $service_group = 'dirsrv'
+) {
+  assert_private()
+
+  $_instance_name = split($title, /^(dirsrv@)?slapd-/)[-1]
+
+  $_default_dse_config = {
+    'cn=encryption,cn=config' => {
+      'allowWeakCipher'               => 'off',
+      'allowWeakDHParam'              => 'off',
+      'nsSSL2'                        => 'off',
+      'nsSSL3'                        => 'off',
+      'nsSSLClientAuth'               => 'required',
+      'nsTLS1'                        => 'on',
+      'nsTLSAllowClientRenegotiation' => 'on',
+      'sslVersionMax'                 => 'TLS1.2',
+      'sslVersionMin'                 => 'TLS1.2'
+    },
+    'cn=config'               => {
+      'nsslapd-ssl-check-hostname' => 'on',
+      'nsslapd-validate-cert'      => 'on',
+      'nsslapd-minssf'             => 256
+    }
+  }
+
+  $_required_dse_config = {
+    'cn=config' => {
+      'nsslapd-security'   => 'on',
+      'nsslapd-securePort' => $port
+    }
+  }
+
+  if $ensure == 'disabled' {
+    ds389::instance::attr::set { "Do not require encryption for ${_instance_name}":
+      instance_name => $_instance_name,
+      root_dn       => $root_dn,
+      root_pw_file  => $root_pw_file,
+      force_ldapi   => true,
+      key           => 'nsslapd-minssf',
+      value         => '0'
+    }
+  }
+  else {
+    # Check to make sure we're not going to have a conflict with something that's running
+    pick($facts['ds389__instances'], {}).each |$daemon, $data| {
```

Comment:
This block of logic for checking a port ('secure_port' in this case), looks nearly identical to the logic in https://github.com/trevor-vaughan/pupmod-simp-ds389/blob/SIMP-8340/manifests/instance.pp#L106-L116.  Looks like it could be put into a function.